air/src/stack/stack_manipulation/mod.rs

```diff
@@ -67,12 +180,176 @@ pub fn enforce_swap_constraints<E: FieldElement>(
+    // frame.
+    result[1] = op_flag.movdn3() * are_equal(frame.stack_item(0), frame.stack_item_next(3));
+
+    // Enforces the top element in the current frame is equal to the 5th element in the next
+    // frame.
+    result[2] = op_flag.movdn4() * are_equal(frame.stack_item(0), frame.stack_item_next(4));
+
+    // Enforces the top element in the current frame is equal to the 6th element in the next
+    // frame.
+    result[3] = op_flag.movdn5() * are_equal(frame.stack_item(0), frame.stack_item_next(5));
+
+    // Enforces the top element in the current frame is equal to the 7th element in the next
+    // frame.
+    result[4] = op_flag.movdn6() * are_equal(frame.stack_item(0), frame.stack_item_next(6));
+
+    // Enforces the top element in the current frame is equal to the 8th element in the next
+    // frame.
+    result[5] = op_flag.movdn7() * are_equal(frame.stack_item(0), frame.stack_item_next(7));
+
+    // Enforces the top element in the current frame is equal to the 9th element in the next
+    // frame.
+    result[6] = op_flag.movdn8() * are_equal(frame.stack_item(0), frame.stack_item_next(8));
+
+    7
+}
+
+/// Enforces constraints of the CSWAP operation. The CSWAP operation swaps the elements
+/// 1,2 in the stack if the first element is 1. The stack remains the same if the top element
+/// is 0. Therefore, the following constraints are enforced:
+/// - The top element should be a binary. s0**2 - s0 = 0.
+/// - The second element in the current frame should be equal to the second element in the
+///   next frame and first element in the next frame should be equal to the third element
+///   in the current frame if the first element in the current frame is a 1.
+/// - The second element in the current frame should be equal to the first element in the
+///   next frame and second element in the next frame should be equal to the third element
+///   in the current frame if the first element in the current frame is a 0.
+pub fn enforce_cswap_constraints<E: FieldElement>(
+    frame: &EvaluationFrame<E>,
+    result: &mut [E],
+    op_flag: E,
+) -> usize {
+    // Enforces the top element in the current is a binary.
+    result[0] = op_flag * is_binary(frame.stack_item(0));
+
+    // Enforces the elements 1,2 are swapped if the top element in the current frame is a 1.
+    let helper_second_0 = frame.stack_item_next(0) - frame.stack_item(0) * frame.stack_item(2);
+    let helper_first_0 = frame.stack_item(1) * binary_not(frame.stack_item(0));
+    result[1] = op_flag * are_equal(helper_second_0, helper_first_0);
+
+    // Enforces the elements 1,2 are not swapped if the top element in the current frame is a 0.
+    let helper_second_1 = frame.stack_item_next(1) - frame.stack_item(0) * frame.stack_item(1);
+    let helper_first_1 = frame.stack_item(2) * binary_not(frame.stack_item(0));
+    result[2] = op_flag * are_equal(helper_second_1, helper_first_1);
```

Comment:
I found this a bit difficult to follow. Naming variables more descriptively could have helped here, but also, I wonder if constraints could be written slightly differently. For example:

```Rust
let condition = frame.stack_item(0);
let not_condition = binary_not(condition);

let a = frame.stack_item(1);
let b = frame.stack_item(2);

result[1] = op_flag * are_equal(frame.stack_item_next(0), a * condition + b * not_condition)
result[2] = op_flag * are_equal(frame.stack_item_next(1), a * not_condition + b * condition)
```
Assuming I got the logic correctly, I think the above is a bit more readable.